readme.md
Method Cacheable
============

Cache method calls and speed up your Ruby on Rails application with MethodCacheable. It's kindof like [ cache_method](https://github.com/seamusabshere/cache_method?utm_source=rubyweekly&utm_medium=email) but it's more explicit about what's being cached and how.


## Simplicity Rules

This is a very very small library wrapped around `Rails.cache` api. It's goal is to be easy to use, and flexible (can be used without Rails). Currently method_cacheable weights in at about 200 lines with documentation (not counting tests and readme)


In your Model include `MethodCacheable`

app/models/user.rb

```ruby

  class User < ActiveRecord::Base
    include MethodCacheable

    has_many :pictures

    def expensive_method(val)
      sleep 120
      return val
    end
  end

```

Then use the `#cache` method to fetch results from cache when available

```
  user = User.last

  # Call User#expensive_method normally
  Benchmark.measure { user.expensive_method(22) }.real
  # => 120.00037693977356

  # Fetch User#expensive_method from cache
  Benchmark.measure { user.cache.expensive_method(22) }.real
  # => 0.000840902328491211
```


Install
=======
in your Gemfile

    gem 'method_cacheable'


You will also want to have a library for caching objects. Such as `dalli` for using memcache

    gem 'dalli'

Then run

    bundle install


## Set up Memcache

In your `app/config/application.rb` set your Rails cache store to use `dalli`

```ruby
  config.cache_store = :dalli_store
```

In an initializer tell MethodCacheable to use the Rails.cache backend. You can use any object here that responds to `#write`, `#read`, and `#fetch`.

initializers/method_cacheable.rb

```ruby
  MethodCacheable.config do |config|
    config.store = Rails.cache
  end
```

then in your models

    include MethodCacheable

Now you're good to go, just use the `cache` method in that class and you can write, read and fetch any method from cache.


Usage
========

By default the `cache` method will will `:fetch` from the cache store. This means that if the key exists it will be pulled, if not the method will be called, returned, and the key will be set.

``` ruby
  user.cache(:read).expensive_method("w00t") # => nil
  user.cache.expensive_method("w00t")        # => "w00t" # sets the cache via :fetch
  user.cache.expensive_method("w00t")        # => "w00t" # pulls from the cache
```

You can also call `:fetch` explicitly if you prefer

``` ruby
  user.cache(:fetch).expensive_method("w00t")  # => "w00t" # pulls from the cache
```


Explicitly write & read methods.

``` ruby
  user.cache(:read).pictures  # => nil
  user.cache(:write).pictures # => [<# Picture ...>, <# Picture ...>] # refreshes the cache
  user.cache(:read).pictures  # => [<# Picture ...>, <# Picture ...>]
```


Different method arguments to the method generate different cache objects. I.E. different input => different output, same input => same output

``` ruby
  user.cache.expensive_method(:schneems => :is_awesome).inspect
  # => {:schneems => :is_awesome}
  user.cache.expensive_method("j/k lol").inspect
  # => "j/k lol"
```

## Delete An Entry

You can delete any cached method using `cache.delete` and passing in the name of the method and any arguements, for example:

```ruby
user.cache.expensive_method("w00t")           # => "w00t"

user.cache.delete(:expensive_method, "w00t")  # => nil
```


Configuration
=============

Any configuration options passed to the cache method will be passed to the cache store (default is [Rails.cache](http://api.rubyonrails.org/classes/ActionController/Caching.html#method-i-cache Rails.cache))

``` ruby
  user.cache(:write, :expires_in => 5.seconds).pictures # => [<# Picture ...>, #... ]
  user.cache(:read).pictures                            # => [<# Picture ...>, #... ]
  sleep 10                                              # => 2
  user.cache(:read).pictures                            # => nil
```

## Generate Keys

You don't need to generate keys, we do that for you using a library called [keytar](http://github.com/schneems/keytar). If you want to see a key you can call key and pass in the name of the method and any arguments into it.


```ruby
  User.find(9).cache.key(:foo) # => "users:foo:9"
```

Contribution
============

Fork away. If you want to chat about a feature idea, or a question you can find me on the twitters [@schneems](http://twitter.com/schneems).  Put any major changes into feature branches. Make sure all tests stay green, and make sure your changes are covered.


licensed under MIT License
Copyright (c) 2011 Schneems. See LICENSE.txt for
further details.



CHANGELOG.md
## 0.1.0

- Fix #3 properly use args if tmp_args not passed

spec/spec_helper.rb
require 'rubygems'


$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '../..', 'lib'))

## Fake rails for testing Rails.cache
class Rails
  def self.cache
    self
  end

  # not quite the same as deleting but it will do
  def self.delete(key, options = {})
    eval("@#{key.gsub(':', '_')} = nil")
  end

  def self.exist?(key)
    !eval("@#{key.gsub(':', '_')}").nil?
  end

  def self.fetch(key, options = {}, &block)
    eval("@#{key.gsub(':', '_')} ||= block.call")
  end

  def self.write(key, val, options = {}, &block)
    eval("@#{key.gsub(':', '_')} = val")
  end

  def self.read(key, options = {})
    eval("@#{key.gsub(':', '_')}")
  end
end

require 'method_cacheable'


MethodCacheable.config do |config|
  config.store = Rails.cache
end


class User
  include MethodCacheable
  define_keys :foo

  def foo(var=nil)
    "bar#{var}"
  end

  def id
   @id ||= rand(100)
  end
end


require 'rspec'
require 'rspec/autorun'



spec/johnny_cache_spec.rb
require 'spec_helper'


describe JohnnyCache do
  before(:each) do
    @user = User.new
    @uniq ||= 0
    @uniq += 1
  end

  describe '' do
    describe 'calling a cached method' do
      describe 'fetch' do
        it 'should return the result of the normal method' do
          @user.cache.foo(@uniq).should == @user.foo(@uniq)
        end

        it 'should bypass the normal method if the cache is written' do
          @user.cache(:write).foo(@uniq)
          @user.should_not_receive(:foo)
          @user.cache(:fetch).foo(@uniq)
        end

        it 'should bypass the normal method if the cache has been fetched before' do
          @user.cache(:fetch).foo(@uniq)
          @user.should_not_receive(:foo)
          @user.cache(:fetch).foo(@uniq)
        end

        it 'should call the normal method if the cache has been not fetched before' do
          @user.should_receive(:foo)
          @user.cache(:fetch).foo(@uniq)
        end

        it 'should call the normal method if the cache has been not fetched before' do
          @user.should_receive(:foo)
          @user.cache.foo(@uniq)
        end

      end

      describe 'read' do
        it 'read should return nil if cache has not been set yet' do
          @user.cache(:read).foo(@uniq).should eq(nil)
        end

        it 'read should return value if cache has been set' do
          result = @user.cache.foo(@uniq)
          @user.cache(:read).foo(@uniq).should eq(result)
        end
      end
    end
  end

end


spec/method_cachable_spec.rb
require 'spec_helper'


describe MethodCachable do
  before(:each) do
    @user = User.new
    @uniq ||= 0
    @uniq += 1
  end

  describe '' do
    describe 'calling a cached method' do
      describe 'fetch' do
        it 'should return the result of the normal method' do
          @user.cache.foo(@uniq).should == @user.foo(@uniq)
        end

        it 'should bypass the normal method if the cache is written' do
          @user.cache(:write).foo(@uniq)
          @user.should_not_receive(:foo)
          @user.cache(:fetch).foo(@uniq)
        end

        it 'should bypass the normal method if the cache has been fetched before' do
          @user.cache(:fetch).foo(@uniq)
          @user.should_not_receive(:foo)
          @user.cache(:fetch).foo(@uniq)
        end

        it 'should call the normal method if the cache has been not fetched before' do
          @user.should_receive(:foo)
          @user.cache(:fetch).foo(@uniq)
        end

        it 'should call the normal method if the cache has been not fetched before' do
          @user.should_receive(:foo)
          @user.cache.foo(@uniq)
        end

      end

      describe 'read' do
        it 'read should return nil if cache has not been set yet' do
          @user.cache(:read).foo(@uniq).should eq(nil)
        end

        it 'read should return value if cache has been set' do
          result = @user.cache.foo(@uniq)
          @user.cache(:read).foo(@uniq).should eq(result)
        end
      end
    end
  end

end


lib/method_cachable.rb
require 'keytar'
require 'active_support/concern'


# include MethodCachable
#
#
# @example
#   class User < ActiveRecord::Base
#     include MethodCachable
#
#     def expensive_method(val)
#       sleep 120
#       return val
#     end
#   end
#
#   user = User.last
#
#   user.expensive_method(22)
#   # => 22
#   user.cache.expensive_method(22)
#   # => 22
#
#   Benchmark.measure do
#     user.expensive_method(22)
#   end.real
#   # => 120.00037693977356
#
#   Benchmark.measure do
#     user.cache.expensive_method(22)
#   end.real
#   # => 0.000840902328491211
#
#   # SOOOOOOOO FAST!!
#
# @see MethodCachable#cache More info on cache options
module MethodCachable
  extend ActiveSupport::Concern
  STORE = nil || Rails.cache


  # @overload cache
  # @overload cache(options = {})
  # @overload cache(cache_operation = :fetch)
  # @overload cache(cache_operation = :fetch, options = {})
  #   Creates a MethodCache instance that performs the given cache operation on the method it receives
  #   @param [Symbol] cache_operation (:fetch) The method called on the cache (:write, :read, or :fetch)
  #   @param [Hash] options Optional hash that gets passed to the cache store
  #
  #
  # @example
  #   user = User.last
  #   # cache
  #   user.cache.some_method
  #
  #   # cache(options)
  #   user.cache(:expires_in => 1.minutes).some_method
  #
  #   # cache(cache_operation)
  #   user.cache(:fetch).some_method
  #   user.cache(:read).some_method
  #   user.cache(:write).some_method
  #
  #   # cache(cache_operation, options)
  #   user.cache(:write, :expires_in 2.minutes).some_method
  def cache(*args)
    MethodCache.new(self, *args)
  end


  module ClassMethods
    # @see MethodCachable#cache
    def cache(*args)
      MethodCache.new(self, *args)
    end
  end

  included do
    include Keytar
  end

  class MethodCache
    attr_accessor :caller_object, :method, :args, :options, :cache_operation

    def initialize(caller_object, *method_cache_args)
      cache_operation      = method_cache_args.map {|x| x if x.is_a? Symbol }.compact.first
      options           = method_cache_args.map {|x| x if x.is_a? Hash   }.compact.first
      self.cache_operation = cache_operation||:fetch
      self.options      = options
      self.caller_object       = caller_object
    end

    # Uses keytar to create a key based on the method and caller if no method_key exits
    # @see http://github.com/schneems/keytar Keytar, it builds keys
    # @return the key used to set the cache
    # @example
    #   cache = User.find(263619).cache   # => #<MethodCachable::MethodCache ... >
    #   cache.method = "foo"              # => "foo"
    #   cache.key                         # => "users:foo:263619"
    def key
      key_method = "#{method}_key".to_sym
      key = caller_object.send key_method, *args if caller_object.respond_to? key_method
      key ||= caller_object.build_key(:name => method, :args => args)
    end

    # Calls the cache based on the given cache_operation
    # @param [Hash] options Options are passed to the cache store
    # @see http://api.rubyonrails.org/classes/ActionController/Caching.html#method-i-cache Rails.cache documentation
    def call_cache_operation(options = {})
      if cache_operation == :fetch
        MethodCachable::STORE.fetch(key, options) do
          caller_object.send method.to_sym, *args
        end
      elsif cache_operation == :read
        MethodCachable::STORE.read(key, options)
      elsif cache_operation == :write
        val = caller_object.send method.to_sym, *args
        MethodCachable::STORE.write(key, val, options)
      end
    end

    # Methods caught by method_missing are passed to the caller_object and used to :write, :read, or :fetch from the cache
    #
    # @see MethodCachable#cache
    def method_missing(method, *args, &blk)
      if caller_object.respond_to? method
        self.method = method
        self.args = args
        call_cache_operation(options)
      else
        super
      end
    end
  end
end


lib/method_cacheable.rb
require 'keytar'
require 'active_support/concern'


# include MethodCacheable
#
#
# @example
#   class User < ActiveRecord::Base
#     include MethodCacheable
#
#     def expensive_method(val)
#       sleep 120
#       return val
#     end
#   end
#
#   user = User.last
#
#   user.expensive_method(22)
#   # => 22
#   user.cache.expensive_method(22)
#   # => 22
#
#   Benchmark.measure do
#     user.expensive_method(22)
#   end.real
#   # => 120.00037693977356
#
#   Benchmark.measure do
#     user.cache.expensive_method(22)
#   end.real
#   # => 0.000840902328491211
#
#   # SOOOOOOOO FAST!!
#
# @see MethodCacheable#cache More info on cache options
module MethodCacheable
  mattr_accessor :store
  extend ActiveSupport::Concern


  def self.config
    yield self
  end



  # @overload cache
  # @overload cache(options = {})
  # @overload cache(cache_operation = :fetch)
  # @overload cache(cache_operation = :fetch, options = {})
  #   Creates a MethodCache instance that performs the given cache operation on the method it receives
  #   @param [Symbol] cache_operation (:fetch) The method called on the cache (:write, :read, or :fetch)
  #   @param [Hash] options Optional hash that gets passed to the cache store
  #
  #
  # @example
  #   user = User.last
  #   # cache
  #   user.cache.some_method
  #
  #   # cache(options)
  #   user.cache(:expires_in => 1.minutes).some_method
  #
  #   # cache(cache_operation)
  #   user.cache(:fetch).some_method
  #   user.cache(:read).some_method
  #   user.cache(:write).some_method
  #
  #   # cache(cache_operation, options)
  #   user.cache(:write, :expires_in 2.minutes).some_method
  def cache(*args)
    MethodCache.new(self, *args)
  end


  module ClassMethods
    # @see MethodCacheable#cache
    def cache(*args)
      MethodCache.new(self, *args)
    end
  end

  included do
    include Keytar
  end

  class MethodCache
    attr_accessor :caller, :method, :args, :options, :cache_operation

    def args=(args)
      args  = [args] unless args.is_a? Array
      @args = args
    end

    def initialize(caller, *method_cache_args)
      self.caller          = caller
      self.cache_operation = method_cache_args.map {|x| x if x.is_a? Symbol }.compact.first||:fetch
      self.options         = method_cache_args.map {|x| x if x.is_a? Hash   }.compact.first
    end

    # Calls the cache based on the given cache_operation
    # @param [Hash] options Options are passed to the cache store
    # @see http://api.rubyonrails.org/classes/ActionController/Caching.html#method-i-cache Rails.cache documentation
    def call
      case cache_operation
      when :fetch
        MethodCacheable.store.fetch(key, options) do
          send_to_caller
        end
      when :read
        read(method, args)
      when :write
        write(method, args) do
          send_to_caller
        end
      end
    end

    def send_to_caller
      caller.send method.to_sym, *args
    end

    def write(method, *args, &block)
      val = block.call
      MethodCacheable.store.write(key, val, options)
    end

    def read(method, *args)
      MethodCacheable.store.read(key, options)
    end

    def for(method , *args)
      self.method = method
      self.args   = args
      self
    end

    # Uses keytar to create a key based on the method and caller if no method_key exits
    # @see http://github.com/schneems/keytar Keytar, it builds keys
    #
    # Method and arguement can optionally be supplied
    #
    # @return the key used to set the cache
    # @example
    #   cache = User.find(263619).cache   # => #<MethodCacheable::MethodCache ... >
    #   cache.method = "foo"              # => "foo"
    #   cache.key                         # => "users:foo:263619"
    def key(tmp_method = nil, *tmp_args)
      tmp_method  = method if tmp_method.blank?
      tmp_args    = args   if tmp_args.blank?
      key_method = "#{tmp_method}_key".to_sym
      key = caller.send key_method, *tmp_args if caller.respond_to? key_method
      key ||= caller.build_key(:name => tmp_method, :args => tmp_args)
    end

    # Removes the current key from the cache store
    def delete(method, *args)
      self.method = method
      self.args   = args
      MethodCacheable.store.delete(key, options)
    end

    # Checks to see if the key exists in the cache store
    # @return [boolean]
    def exist?(method, *args)
      self.method = method
      self.args   = args
      MethodCacheable.store.exist?(key)
    end
    alias :exists? :exist?

    # Methods caught by method_missing are passed to the caller and used to :write, :read, or :fetch from the cache
    #
    # @see MethodCacheable#cache
    def method_missing(method, *args, &blk)
      self.method = method
      self.args   = args
      if caller.respond_to? method
        call
      else
        send_to_caller
      end
    end
  end
end


lib/johnny_cache.rb
require 'rubygems' #todo remove

require 'keytar'
require 'active_support/concern'


# include JohnnyCache
#
#
# @example
#   class User < ActiveRecord::Base
#     include JohnnyCache
#
#     def expensive_method(val)
#       sleep 120
#       return val
#     end
#   end
#
#   user = User.last
#
#   user.expensive_method(22)
#   # => 22
#   user.cache.expensive_method(22)
#   # => 22
#
#   Benchmark.measure do
#     user.expensive_method(22)
#   end.real
#   # => 120.00037693977356
#
#   Benchmark.measure do
#     user.cache.expensive_method(22)
#   end.real
#   # => 0.000840902328491211
#
#   # SOOOOOOOO FAST!!
#
# @see JohnnyCache#cache More info on cache options
module JohnnyCache
  extend ActiveSupport::Concern
  STORE = nil || Rails.cache


  # @overload cache
  # @overload cache(options = {})
  # @overload cache(cache_operation = :fetch)
  # @overload cache(cache_operation = :fetch, options = {})
  #   Creates a MethodCache instance that performs the given cache operation on the method it receives
  #   @param [Symbol] cache_operation (:fetch) The method called on the cache (:write, :read, or :fetch)
  #   @param [Hash] options Optional hash that gets passed to the cache store
  #
  #
  # @example
  #   user = User.last
  #   # cache
  #   user.cache.some_method
  #
  #   # cache(options)
  #   user.cache(:expires_in => 1.minutes).some_method
  #
  #   # cache(cache_operation)
  #   user.cache(:fetch).some_method
  #   user.cache(:read).some_method
  #   user.cache(:write).some_method
  #
  #   # cache(cache_operation, options)
  #   user.cache(:write, :expires_in 2.minutes).some_method
  def cache(*args)
    MethodCache.new(self, *args)
  end


  module ClassMethods
    # @see JohnnyCache#cache
    def cache(*args)
      MethodCache.new(self, *args)
    end
  end

  included do
    include Keytar
  end

  class MethodCache
    attr_accessor :caller_object, :method, :args, :options, :cache_operation

    def initialize(caller_object, *method_cache_args)
      cache_operation      = method_cache_args.map {|x| x if x.is_a? Symbol }.compact.first
      options           = method_cache_args.map {|x| x if x.is_a? Hash   }.compact.first
      self.cache_operation = cache_operation||:fetch
      self.options      = options
      self.caller_object       = caller_object
    end

    # Uses keytar to create a key based on the method and caller if no method_key exits
    # @see http://github.com/schneems/keytar Keytar, it builds keys
    # @return the key used to set the cache
    # @example
    #   cache = User.find(263619).cache   # => #<JohnnyCache::MethodCache ... >
    #   cache.method = "foo"              # => "foo"
    #   cache.key                         # => "users:foo:263619"
    def key
      key_method = "#{method}_key".to_sym
      key = caller_object.send key_method, *args if caller_object.respond_to? key_method
      key ||= caller_object.build_key(:name => method, :args => args)
    end

    # Calls the cache based on the given cache_operation
    # @param [Hash] options Options are passed to the cache store
    # @see http://api.rubyonrails.org/classes/ActionController/Caching.html#method-i-cache Rails.cache documentation
    def call_cache_operation(options = {})
      if cache_operation == :fetch
        JohnnyCache::STORE.fetch(key, options) do
          caller_object.send method.to_sym, *args
        end
      elsif cache_operation == :read
        JohnnyCache::STORE.read(key, options)
      elsif cache_operation == :write
        val = caller_object.send method.to_sym, *args
        JohnnyCache::STORE.write(key, val, options)
      end
    end

    # Methods caught by method_missing are passed to the caller_object and used to :write, :read, or :fetch from the cache
    #
    # @see JohnnyCache#cache
    def method_missing(method, *args, &blk)
      if caller_object.respond_to? method
        self.method = method
        self.args = args
        call_cache_operation(options)
      else
        super
      end
    end
  end
end
